StackOverflowError when running dokkaHtml task

**Describe the bug**
`StackOverflowError` occurs when running `dokkaHtml` Gradle task.

**Expected behaviour**
Documentation to be built sucessfully

**Screenshots**
None

**To Reproduce**
- Run `dokkaHtml` task. 

I'm able to reproduce this bug on [this project](https://github.com/blogify-dev/krate/tree/1.4-update) on commit https://github.com/blogify-dev/krate/commit/200bb00fb76fe9f85aa211948a800a66cbfe83ea
Traceback: https://pastify-app.web.app/show/e4A2CI6lXOh1QcQXMBaR

**Dokka configuration**
Configuration of dokka used to reproduce the bug

```kotlin
    dokkaHtml {
        outputDirectory = "$buildDir/dokka"

        dokkaSourceSets {
            configureEach {
                includeNonPublic = false

                skipDeprecated = true

                skipEmptyPackages = true

                jdkVersion = 8

                includes = listOf("docs/krate.md")

                externalDocumentationLink {
                    url = URL("https://docs.31416.dev/reflectr/")
                }

                perPackageOption {
                    prefix   = "krate.annotations"
                    suppress = true
                }

                perPackageOption {
                    prefix   = "krate.util"
                    suppress = true
                }

                sourceLink {
                    path = "./"

                    url = "https://github.com/blogify-dev/krate/blob/master/"

                    lineSuffix = "#L"
                }
            }
        }
    }
```

This configuration in the project used can be found [here](https://github.com/blogify-dev/krate/blob/200bb00fb76fe9f85aa211948a800a66cbfe83ea/build.gradle.kts#L114)

**Installation**
- Operating system: Linux (Fedora 32)
- Build tool: Gradle v6.5
- Dokka version: 1.4.0-rc

**Additional context**
None. If I missed anything or need to provide more information, please let me know.

@Kordyjan 
Here is the prepared project you can use for debugging ☺️ 
https://github.com/sellmair/krate/tree/dokka-bug-1342

<Details>
<summary>Here is the Stacktrace</summary>
```

> Task :dokkaHtml FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dokkaHtml'.
> java.lang.StackOverflowError (no error message)

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':dokkaHtml'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:207)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:205)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:186)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:356)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.StackOverflowError
        at org.jetbrains.kotlin.types.ErrorUtils.isError(ErrorUtils.java:547)
        at org.jetbrains.kotlin.resolve.DescriptorUtils.getFqNameSafeIfPossible(DescriptorUtils.java:99)
        at org.jetbrains.kotlin.resolve.DescriptorUtils.getFqNameSafe(DescriptorUtils.java:92)
        at org.jetbrains.kotlin.resolve.descriptorUtil.DescriptorUtilsKt.getFqNameSafe(DescriptorUtils.kt:56)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.getConstructorName(TypeReferenceFactory.kt:51)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:38)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromProjection(TypeReferenceFactory.kt:47)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:39)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.fromPossiblyNullable(TypeReferenceFactory.kt:28)
        at org.jetbrains.dokka.analysis.TypeReferenceFactoryKt.from(TypeReferenceFactory.kt:35)

```
</Details>

--- a/core/src/main/kotlin/links/DRI.kt
+++ b/core/src/main/kotlin/links/DRI.kt
@@ -68,8 +68,8 @@ data class TypeConstructor(
             (if (params.isNotEmpty()) "[${params.joinToString(",")}]" else "")
 }
 
-object SelfType : TypeReference() {
-    override fun toString() = "^"
+data class RecursiveType(val rank: Int): TypeReference() {
+    override fun toString() = "^".repeat(rank + 1)
 }
 
 data class Nullable(val wrapped: TypeReference) : TypeReference() {

--- a/kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/TypeReferenceFactory.kt
+++ b/kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/TypeReferenceFactory.kt
@@ -12,7 +12,7 @@ import org.jetbrains.kotlin.types.TypeProjection
 
 fun TypeReference.Companion.from(d: ReceiverParameterDescriptor): TypeReference? =
     when (d.value) {
-        is ExtensionReceiver -> fromPossiblyNullable(d.type)
+        is ExtensionReceiver -> fromPossiblyNullable(d.type, emptyList())
         else -> run {
             println("Unknown value type for $d")
             null
@@ -20,31 +20,35 @@ fun TypeReference.Companion.from(d: ReceiverParameterDescriptor): TypeReference?
     }
 
 fun TypeReference.Companion.from(d: ValueParameterDescriptor): TypeReference? =
-    fromPossiblyNullable(d.type)
+    fromPossiblyNullable(d.type, emptyList())
 
 fun TypeReference.Companion.from(p: PsiClass) = TypeReference
 
-private fun TypeReference.Companion.fromPossiblyNullable(t: KotlinType, self: KotlinType? = null): TypeReference =
-    from(t, self).let { if (t.isMarkedNullable) Nullable(it) else it }
+private fun TypeReference.Companion.fromPossiblyNullable(t: KotlinType, paramTrace: List<KotlinType>): TypeReference =
+    fromPossiblyRecursive(t, paramTrace).let { if (t.isMarkedNullable) Nullable(it) else it }
 
-private fun TypeReference.Companion.from(t: KotlinType, self: KotlinType? = null): TypeReference =
-    if (self is KotlinType && self.constructor == t.constructor && self.arguments == t.arguments)
-        SelfType
-    else when (val d = t.constructor.declarationDescriptor) {
+private fun TypeReference.Companion.fromPossiblyRecursive(t: KotlinType, paramTrace: List<KotlinType>): TypeReference =
+    paramTrace.indexOfFirst { it.constructor == t.constructor && it.arguments == t.arguments }
+        .takeIf { it >= 0 }
+        ?.let(::RecursiveType)
+        ?: from(t, paramTrace)
+
+private fun TypeReference.Companion.from(t: KotlinType, paramTrace: List<KotlinType>): TypeReference =
+    when (val d = t.constructor.declarationDescriptor) {
         is TypeParameterDescriptor -> TypeParam(
-            d.upperBounds.map { fromPossiblyNullable(it, self ?: t) }
+            d.upperBounds.map { fromPossiblyNullable(it, paramTrace + t) }
         )
         else -> TypeConstructor(
             t.constructorName.orEmpty(),
-            t.arguments.map { fromProjection(it, self) }
+            t.arguments.map { fromProjection(it, paramTrace) }
         )
     }
 
-private fun TypeReference.Companion.fromProjection(t: TypeProjection, r: KotlinType? = null): TypeReference =
+private fun TypeReference.Companion.fromProjection(t: TypeProjection, paramTrace: List<KotlinType>): TypeReference =
     if (t.isStarProjection) {
         StarProjection
     } else {
-        fromPossiblyNullable(t.type, r)
+        fromPossiblyNullable(t.type, paramTrace)
     }
 
 private val KotlinType.constructorName

--- a/plugins/base/src/test/kotlin/basic/DRITest.kt
+++ b/plugins/base/src/test/kotlin/basic/DRITest.kt
@@ -38,7 +38,7 @@ class DRITest : AbstractCoreTest() {
                 val expected = TypeConstructor(
                     "kotlin.Function1", listOf(
                         TypeParam(listOf(Nullable(TypeConstructor("kotlin.Any", emptyList())))),
-                        Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(SelfType)))))
+                        Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(RecursiveType(0))))))
                     )
                 )
                 val actual = module.packages.single()
@@ -70,7 +70,7 @@ class DRITest : AbstractCoreTest() {
             configuration
         ) {
             documentablesMergingStage = { module ->
-                val expected = Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(SelfType)))))
+                val expected = Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(RecursiveType(0))))))
                 val actual = module.packages.single()
                     .functions.single()
                     .dri.callable?.params?.single()
@@ -100,7 +100,7 @@ class DRITest : AbstractCoreTest() {
             configuration
         ) {
             documentablesMergingStage = { module ->
-                val expected = Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(SelfType)))))
+                val expected = Nullable(TypeParam(listOf(TypeConstructor("kotlin.Comparable", listOf(RecursiveType(0))))))
                 val actual = module.packages.single()
                     .functions.single()
                     .dri.callable?.receiver
@@ -311,6 +311,35 @@ class DRITest : AbstractCoreTest() {
                      documentable.generics.first().dri.toString()
                 )
 
+            }
+        }
+    }
+
+    @Test
+    fun `deep recursive typebound #1342`() {
+        val configuration = dokkaConfiguration {
+            sourceSets {
+                sourceSet {
+                    sourceRoots = listOf("src/")
+                }
+            }
+        }
+        testInline(
+            """
+            |/src/main/kotlin/Test.kt
+            |package example
+            |
+            | fun <T, S, R> recursiveBound(t: T, s: S, r: R) where T: List<S>, S: List<R>, R: List<S> = Unit
+            |
+        """.trimMargin(),
+            configuration
+        ) {
+            documentablesMergingStage = { module ->
+                val function = module.dfs { it.name == "recursiveBound" }
+                assertEquals(
+                    "example//recursiveBound/#TypeParam(bounds=[kotlin.collections.List[TypeParam(bounds=[kotlin.collections.List[TypeParam(bounds=[kotlin.collections.List[^^]])]])]])#TypeParam(bounds=[kotlin.collections.List[TypeParam(bounds=[kotlin.collections.List[^]])]])#TypeParam(bounds=[kotlin.collections.List[TypeParam(bounds=[kotlin.collections.List[^]])]])/PointingToDeclaration/",
+                    function?.dri?.toString(),
+                )
             }
         }
     }

